Project Proposal: Feature Flag Semantic Conventions (#2042)
open-telemetry/community: projects/feature-flag.md

- Identify which services are evaluating a specific feature flag.
  This is useful for the process of removing flags from your code and feature flagging service, an important part of the flag lifecycle.
- Analyzing the impact of the actual flag evaluation.
  Some feature flag libraries and services do async work such as database or remote calls to decide which variant of a feature flag should be returned.

## Deliverables

The project deliverable will be a stable feature flag semantic convention for flag evaluations, also called impressions, and feature flag changes.
Both client and server uses should be considered, and prototypes should be created for at least one server and at least one client use case.

While prototypes will be delivered as a part of the development effort of this project, stable instrumentation is out of scope of this project and is expected to come from the OpenFeature project in the form of OpenFeature hooks or native instrumentation in OpenFeature SDKs.
Any use of baggage, including propagating fields to be used in feature flag evaluation context, is also out of scope.

> ## Staffing

Reviewer: I can at least help to review prototypes.

Ryan Lamb (OpenFeature TC, LaunchDarkly)

Author: Added you to the list

# Feature Flag Semantic Conventions

> OpenFeature is a CNCF Incubating project aiming to standardize feature flags.

Reviewer 2: ```suggestion
[OpenFeature](https://openfeature.dev/) is a CNCF Incubating project aiming to standardize feature flags.
```

They currently provide 11 open source SDKs in various languages and support more than 10 feature flagging backends.

> OpenFeature has had contributions from over 50 companies and is actively developed by a combination of end users, feature flagging vendors, and platform providers.

Reviewer 2: Could link to the lists at https://openfeature.dev/community/interested-parties.

OpenFeature provides observability data via OpenTelemetry hooks.
A hook is an OpenFeature plugin which may execute logic at various stages of the feature flag evaluation process.

> Currently, those hooks use the experimental feature flagging semantic convention.

Reviewer 2: ```suggestion
Currently, those hooks use the experimental [feature flagging semantic convention](https://opentelemetry.io/docs/specs/semconv/feature-flags/).
```

Any use of baggage, including propagating fields to be used in feature flag evaluation context, is also out of scope.

## Staffing

The project is expected to be mainly staffed by OpenFeature contributors representing a strong subsection of the feature flagging vendor and user community.
This group will provide the domain expertise to ensure the feature flag observability meets the requirements of feature flag vendors and users.

This project also includes Dan Dyla acting as one of the project leads.
Dan is a member of the GC, maintainer of OTel JS, and has experience developing semantic conventions.
Dan will provide the necessary expertise to ensure the proposed semantic conventions meet the standards and expectations of the OpenTelemetry community and semantic conventions SIG.

The following is a list of engineers committed to working on the project:

- Daniel Dyla (OTel GC, OTel JS, Dynatrace)
- Ryan Lamb (OpenFeature TC, LaunchDarkly)

> - TODO: more engineers

Reviewer 2: Should this be moved/merged into `### Required staffing`?

This group will provide the domain expertise to ensure the feature flag observability meets the requirements of feature flag vendors and users.

This project also includes Dan Dyla acting as one of the project leads.
Dan is a member of the GC, maintainer of OTel JS, and has experience developing semantic conventions.
Dan will provide the necessary expertise to ensure the proposed semantic conventions meet the standards and expectations of the OpenTelemetry community and semantic conventions SIG.

The project is sponsored by Dan Dyla and Alexander Wert, an OpenTelemetry semantic conventions maintainer.

### Required staffing

**Project Leads:** Daniel Dyla and Michael Beemer (OpenFeature co-founder/GC member)

**Project Sponsors**: Daniel Dyla and Alexander Wert

**Engineers willing to write and review prototypes:**

TODO: list prototype languages and engineers

The following is a list of engineers committed to working on the project:

- JS Prototype - Daniel Dyla (OTel GC, OTel JS, Dynatrace)
- Review Prototypes - Ryan Lamb (OpenFeature TC, LaunchDarkly)

> - TODO: more engineers

Reviewer 3: I can help with prototypes / review / testing.

- Federico Bond (OpenFeature Python, Independent)

> - TODO: more engineers

Reviewer 4: ```suggestion
- Federico Bond (OpenFeature Python, Independent)
- TODO: more engineers
```

## Labels

This project will use the label [`area:feature-flag`](https://github.com/open-telemetry/semantic-conventions/labels/area%3Afeature-flag) to track issues and pull requests in the semantic conventions repository.

## Project Board

Once the project is approved a project board will be created and linked here.
The project lead and all other relevant project members should have edit access to the board.
The board will be pre-populated with at least the following issues to track deliverables and the timeline.

- **signal type for feature flag impressions** - agree on which signal type or types should be used to represent feature flag impressions.
- **signal type for feature flag changes** - agree on which signal type or types should be used to represent feature flag changes.
- **semantic conventions for feature flag impressions** - make any necessary additions to the semantic conventions to support feature flag impressions
- **semantic conventions for feature flag changes** - make any necessary additions to the semantic conventions to support feature flag change events
- **prototype feature flag impressions**

> - **prototype feature flag change events** - feature flag change events may not be generated by SDKs or instrumentations, but by the feature flag management tools.

Reviewer 5: Just a minor comment on this particular point here.

If the management tooling raises the event, this could create a temporal inconsistency where some or all of the applications have not yet fetched the updated values.

In the .NET ecosystem (I'm speaking only on the platform I have experience with here, but this might apply to others), it is incredibly common for one to set a new value to a key (for example, on Azure AppConfiguration), but then have many applications waiting on different refresh timings or cache expirations for that key, meaning they would potentially still serve N requests using the previous value. Even when using something like a per-request middleware, the refresh operation tends to happen in the background and will usually not affect the _current_ request or other parallel requests happening around the same time.

If the only source for a flag change events is the central tool, one could end up checking behavior still running on the old value while the metric/etc indicates a new value should be present.

I think this needs to be reconsidered from the perspective of each application/process as well. Maybe we need a change event from the configuration source, and a second change event on each application when the new value is fetched? That to me would be the most comprehensive approach, although I'm not sure exactly how those event streams would be represented.

I almost want to suggest that OpenTelemetry should define a 4th type of telemetry called "events" (or "state transitions" or something), generalizing the concept and then making span events a particular use case of that. Then, when looking at a particular trace, one could include "global events" like these to understand the complete behavior on a given flow.

Anyways... just throwing some ideas around here.

Reviewer 6: We planned to define semantics for change events from the management tool (source of truth) and the services (SDKs). That way, you could track how long it took for a change to propagate and quickly identify which services haven't been updated.

OpenTelemetry already has an [events concept](https://opentelemetry.io/docs/specs/otel/logs/event-api/) that's basically semantics on top of logs.

Reviewer 5: > We planned to define semantics for change events from the management tool (source of truth) and the services (SDKs). That way, you could track how long it took for a change to propagate and quickly identify which services haven't been updated.

Perfect. Love to hear it.

> OpenTelemetry already has an [events concept](https://opentelemetry.io/docs/specs/otel/logs/event-api/) that's basically semantics on top of logs.

I remember seeing this but always assumed it was just a "synonym" to logs. Looks like I was wrong as it even has its own classes and abstractions.

Am I correct in that observability tools don't quite have this separation between logs and events yet though? Should they? We use Datadog extensively and I never saw this distinction anywhere over there (although I'm not necessarily an expert in DD either... so it could be me just missing it). 

I also don't know what exactly maps to these events from the .NET APIs, since we only have `ILogger` (for logs), `Meter` and `Instrument`s (for metrics) and `Activity` (for spans) there. The only place where I ever saw the concept of "events" was for "activity events", and those only make sense if you _are_ under an activity in the first place.

Author: >I remember seeing this but always assumed it was just a "synonym" to logs. Looks like I was wrong as it even has its own classes and abstractions.

Events and logs are the same data type. Events are LogRecords which have a name which uniquely defines a particular class or type of event. All events with the same name have Payloads that conform to the same schema, which assists in analysis in observability platforms. Events are described in more detail in the [semantic conventions](https://opentelemetry.io/docs/specs/semconv/general/events/).

>Am I correct in that observability tools don't quite have this separation between logs and events yet though? Should they?

For the most part they don't. There could be arguments made either way as to if they should.

Reviewer 7: Hello! I'm part of the `Microsoft.FeatureManagement` team.

> If the only source for a flag change events is the central tool, one could end up checking behavior still running on the old value while the metric/etc indicates a new value should be present.

We solve this with versioning emitted on the `feature_flag` event. For us- the events usually come from the SDK, but the centralized server dictates what version we're working with. This allows a given user/account/activity to be connected to the appropriate version of the flag, meaning you know which version applies to the rest of the metrics.

Reviewer 7: Sadly, I'm unable to join at this time. Very interested to see what y'all land on.

This group will provide the domain expertise to ensure the feature flag observability meets the requirements of feature flag vendors and users.

This project also includes Dan Dyla acting as one of the project leads.
Dan is a member of the GC, maintainer of OTel JS, and has experience developing semantic conventions.
Dan will provide the necessary expertise to ensure the proposed semantic conventions meet the standards and expectations of the OpenTelemetry community and semantic conventions SIG.

The project is sponsored by Dan Dyla and Alexander Wert, an OpenTelemetry semantic conventions maintainer.

### Required staffing

**Project Leads:** Daniel Dyla and Michael Beemer (OpenFeature co-founder/GC member)

**Project Sponsors**: Daniel Dyla and Alexander Wert

**Engineers willing to write and review prototypes:**

TODO: list prototype languages and engineers

The following is a list of engineers committed to working on the project:

- JS Prototype - Daniel Dyla (OTel GC, OTel JS, Dynatrace)
- Review Prototypes - Ryan Lamb (OpenFeature TC, LaunchDarkly)
- Federico Bond (OpenFeature Python, Independent)

> - TODO: more engineers

Reviewer 8: ```suggestion
- André Silva (OpenFeature dotnet)
- TODO: more engineers
```

From @beeme1mr comment: https://github.com/open-telemetry/community/pull/2042#issuecomment-2099260443

OpenFeature provides observability data via OpenTelemetry hooks.
A hook is an OpenFeature plugin which may execute logic at various stages of the feature flag evaluation process.
Currently, those hooks use the experimental [feature flagging semantic convention](https://opentelemetry.io/docs/specs/semconv/feature-flags/).

OpenFeature is looking to expand their observability solution either by improving the currently available hooks or by providing native instrumentation within the OpenFeature SDK itself.
In order to support that effort, OpenFeature requires a stable semantic convention to build instrumentation around.
This project aims to develop and stabilize feature flag semantic conventions, using the existing experimental semantic conventions as a starting point.

## Problem 1: Determining Flag Impact

Currently, it is impossible to determine the impact a feature flag has on key metrics for your services.
For example, if a feature flag is enabled for 25% of users, additional data is required to know if that subset of users is experiencing faster, slower, or unchanged response times.

> This problem can become exponentially more challenging if multiple feature flags are evaluated over the course of a single transaction.

Reviewer 9: ```suggestion
This problem can become exponentially more challenging if multiple feature flags are evaluated over the course of a single trace.
```

Further, it may also be difficult or impossible to know exactly how many users are being affected by a particular feature flag.

> For example, if a feature flag targets users with a particular email domain, it may require additional monitoring data to determine what portion of your users or traffic is receiving a particular experience.

Reviewer 9: ```suggestion
For example, if a feature flag targets users with a particular email domain, it may require additional observability data to determine what portion of your users or traffic is receiving a particular experience.
```

> ## Problem 2: Flag management separate from monitoring data
> 
> Feature flags are typically managed outside your monitoring tooling.
> There is no standardized way for a flag management system to notify a monitoring system that a feature flag rule set is changed, or if a rule is enabled or disabled.
> This lack of standardization means that even if a flag management system emits a flag change event, the monitoring tool may not be able to correlate that change with the services that it impacts.
> It also means that a person responsible for monitoring an application often has to check two separate tools in order to determine if a flag change was the root cause of a change in service metrics or behavior.
> Because rule sets and monitoring data are managed in separate tools, it may be difficult to determine exactly which rule set was used and if it has been changed between when it was evaluated and when a person investigating an issue is looking at the monitoring data.

Reviewer 9: ```suggestion
## Problem 2: Flag management separate from observability data

Feature flags are typically managed outside your observability tooling.
There is no standardized way for a flag management system to notify a observability system that a feature flag rule set is changed, or if a rule is enabled or disabled.
This lack of standardization means that even if a flag management system emits a flag change event, the observability tool may not be able to correlate that change with the services that it impacts.
It also means that a person responsible for the observability of an application often has to check two separate tools in order to determine if a flag change was the root cause of a change in service metrics or behavior.
Because rule sets and observability data are managed in separate tools, it may be difficult to determine exactly which rule set was used and if it has been changed between when it was evaluated and when a person investigating an issue is looking at the observability data.
```

## Target Use Cases

- Analyze the impact a feature flag has on key service metrics such as response time and failure rate.

>   If a feature variant is enabled for some subset of users, it is necessary to be able to distinguish those transactions from other transactions in your monitoring backend.

Reviewer 9: ```suggestion
  If a feature variant is enabled for some subset of users, it is necessary to be able to distinguish those traces from other traces in your observability backend.
```

The project is expected to be mainly staffed by OpenFeature contributors representing a strong subsection of the feature flagging vendor and user community.
This group will provide the domain expertise to ensure the feature flag observability meets the requirements of feature flag vendors and users.

This project also includes Dan Dyla acting as one of the project leads.
Dan is a member of the GC, maintainer of OTel JS, and has experience developing semantic conventions.
Dan will provide the necessary expertise to ensure the proposed semantic conventions meet the standards and expectations of the OpenTelemetry community and semantic conventions SIG.

The project is sponsored by Dan Dyla and Alexander Wert, an OpenTelemetry semantic conventions maintainer.

### Required staffing

**Project Leads and Sponsors:**

- Alexander Wert - Project Sponsor (OTel SemConv, Elastic)
- Daniel Dyla - Project Lead and Sponsor (OTel GC, OTel JS, Dynatrace)
- Michael Beemer - Project Lead (OpenFeature co-founder and GC, Dynatrace)

**Engineers:**

The following is a list of engineers committed to working on the project and write and review prototypes:

- Alexander Wert - (OTel SemConv, Elastic)

> - André Silva (OpenFeature dotnet)

Reviewer 8: ```suggestion
- André Silva (OpenFeature dotnet, Lexis Nexis Risk Solutions)
```

### Required staffing

**Project Leads and Sponsors:**

- Alexander Wert - Project Sponsor (OTel SemConv, Elastic)
- Daniel Dyla - Project Lead and Sponsor (OTel GC, OTel JS, Dynatrace)
- Michael Beemer - Project Lead (OpenFeature co-founder and GC, Dynatrace)

**Engineers:**

The following is a list of engineers committed to working on the project and write and review prototypes:

- Alexander Wert - (OTel SemConv, Elastic)
- André Silva (OpenFeature dotnet)
- Daniel Dyla - (OTel GC, OTel JS, Dynatrace)
- Federico Bond (OpenFeature Python, Independent)

> - Lucas Reining (OpenFeature TC)

Reviewer 10: My name is Lukas with `k` @dyladan :) and maybe you can add `codecentric` as company.
